[CDAP-17241] Add Authorization Enforcement for Starting Preview Runs

dli357 opened the pull request:
Currently, preview does not perform any authorization and, instead, sends all preview run requests on behalf of the master principal. This change adds authorization enforcement for DefaultPreviewManager endpoints. See [CDAP-17241](https://cdap.atlassian.net/browse/CDAP-17241) for additional info.

In order to support using the AuthEnforce annotation, we needed to add parent support for it, so also went ahead and implemented parent support for the AuthEnforce annotation.

Tested using sandbox.

chtyim commented:
If you only enforcing on namespace, isn't that you can do the same for other preview endpoints?

dli357 commented:
@chtyim  Yeah you're right, if our auth extension only authorizes on namespace, it will work. However, other auth extensions which authorize on finer granularity than namespace will fail.

For a longer-term solution, I was thinking upon starting a preview run, the preview manager makes a call to the authorization extension to add an allow policy for the applicationId that was just created, and then enforcement along other endpoints will work.

EDIT: This is an outdated comment, see the updated description and comments below.

dli357 committed:
[CDAP-17241] Added parent enforcement support for AuthEnforce annotation and authorization enforcement for preview

Modified Action.ADMIN javadoc to more accurately reflect its usage

Modified copyright year

dli357 commented:
Modified to add namespace-level enforcement to all the DefaultPreviewManager endpoints. Please review again @chtyim, thanks!

wyzhang commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewManager.java:
```diff
@@ -199,7 +201,8 @@ protected void shutDown() throws Exception {
 
   @Override
-  public ApplicationId start(NamespaceId namespace, AppRequest<?> appRequest) throws Exception {
+  @AuthEnforce(entities = "namespaceId", enforceOn = NamespaceId.class, actions = Action.ADMIN)
```
Prefer read and write instead of admin. But feel free to leave this as is for now. 

Admin has additional pipeline execution permission that we may or may not give to pipeline editor during their testing phase.

wyzhang replied:
As discussed offline, Action.Admin is more like  system/platform read + write permissions

So it is okay to require such Admin permission for preview. In the future, might be better to break this up into separate permissions.

dli357 replied:
Acknowledged. As you stated, READ, WRITE, and EXECUTE are all their own actions, whereas ADMIN currently is more of its own "platform READ/WRITE" privilege rather than a combination of read, write, and execute. We definitely should look into splitting this action up in the future to allow for finer-grain permission checks. I went ahead and modified the Javadoc for Action.ADMIN to more accurately reflect what it does.
